### PR TITLE
Pytorch training updated + LRP

### PR DESCRIPTION
**Pull request by faroukmokhtar**

### CODE REVIEW

File: mlpf/pytorch_delphes/LRP/hooks.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        else:
+            print('Will use single_gpu..')
+except Exception as e:
+    print("Could not import setGPU, running CPU-only")
+
+#define the global base device
+if use_gpu:
+    device = torch.device('cuda:0')
+else:
+    device = torch.device('cpu')
```
Comment (jpata) [inline mlpf/pytorch_delphes/LRP/hooks.py:33]: this seems to be repeated in several places, can you run this GPU detection once from a common file?

File: mlpf/pytorch_delphes/LRP/hooks.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from torch_scatter import scatter_mean
+from torch_geometric.nn.inits import reset
+from torch_geometric.data import Data, DataLoader, DataListLoader, Batch
+from torch_geometric.nn import GravNetConv
+from torch.utils.data import random_split
+import torch_cluster
+
+sys.path.insert(1, '../')
+sys.path.insert(1, '../../../plotting/')
+sys.path.insert(1, '../../../mlpf/plotting/')
```
Comment (jpata) [inline mlpf/pytorch_delphes/LRP/hooks.py:49]: I think this pattern could give some issues with python modules. You should have proper modules defined by placiong `__init__.py` files in the appropriate locations, and be able to do relative imports as in e.g. https://github.com/jpata/particleflow/blob/master/mlpf/tfmodel/model.py#L8

File: mlpf/pytorch_delphes/LRP/hooks.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    x = self.relu(self.conv(x))
+    x.register_hook(lambda grad : torch.clamp(grad, min = 0))     #No gradient shall be backpropagated
+                                                                  #conv outside less than 0
+
+    # print whether there is any negative grad
+    s=x.register_hook(lambda grad: torch.zeros(grad.shape))
+    return self.fc1(self.flatten(x))
+
+
+net = myNet()
```
Comment (jpata) [inline mlpf/pytorch_delphes/LRP/hooks.py:90]: is this a test file (in which case perhaps the file should be moved/removed, and the following code put in `__main__`), or a library file to be imported elsewhere (in which case, this code should be removed)?

File: mlpf/pytorch_delphes/graph_data_delphes.py [hunk 4]
```diff
@@ -7,15 +7,17 @@
 from torch_geometric.data import Dataset, Data, Batch
 import itertools
 from glob import glob
+import numba
```
Comment (jpata) [inline mlpf/pytorch_delphes/graph_data_delphes.py:10]: is numba actually used?

File: mlpf/pytorch_delphes/gravnet.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+except ImportError:
+    knn = None
+
+# copied it from pytorch_geometric source code
+# ADDED: retrieve edge_index, retrieve edge_weight
+# CHANGED: self.lin -> self.lin_p
+# CHANGED: used reduce='sum' instead of reduce='mean' in the message passing
+# REMOVED: skip connection
+
+class GravNetConv(MessagePassing):
```
Comment (jpata) [inline mlpf/pytorch_delphes/gravnet.py:22]: this seems duplicated with respect to gravnet_LRP.py, can you make a common import?

File: scripts/local_test_delphes_pytorch.sh [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+# !/bin/bash
```
Comment (jpata) [inline scripts/local_test_delphes_pytorch.sh:1]: typo?

File: scripts/local_test_delphes_pytorch.sh [hunk 7]
```diff
@@ -10,32 +10,44 @@ mkdir -p data/pythia8_ttbar
...
-COMET_API_KEY="bla" python3 ../mlpf/pytorch/train_end2end_delphes.py \
-  --dataset data/pythia8_ttbar --space_dim 2 --n_train 1 \
-  --n_val 1 --model PFNet7 --convlayer gravnet-radius --convlayer2 "none" \
-  --lr 0.0001 --hidden_dim 32 --n_epochs 3 --l1 1.0 --l2 0.001 --target gen \
-  --batch_size 1 --dropout 0.2 --disable_comet
-
-# predict on some test data and make plots
-python3 ../mlpf/pytorch/eval_end2end_delphes.py --dataset data/pythia8_ttbar \
-  --path experiments/PFNet* --model PFNet7 --start 1 --stop 2 --epoch 1 --target gen
+echo Begining the training..
```
Comment (jpata) [inline scripts/local_test_delphes_pytorch.sh:48]:
```suggestion
echo Beginning the training..
```

**Review by jpata — COMMENTED** [review]
Thanks for the PR! Since we are now a bit more people, let me take the occasion to make some detailed comments to improve the state of the codebase and to exercise good software engineering practices.

1. Could you please provide a PR description with some details what you are changing and implementing here?
1. Could you please comment on the choice of making the pytorch model now specific for the delphes dataset? Can this be made generic, so we could use other datasets as well, as for the TF model?
1. It seems there are a few repeated patterns that could be improved in the code
   1. GPU detection and device choice in a common file, to be imported elsewhere (in the beginning of the main pipeline entry point ideally)
   2. reduce the amount of duplicate definitions for the models (e.g. the GravNet layer)
   2. remove `sys.path.insert` workaround, add `__init__.py` files to make sure module imports can be done cleanly
   3. removing executable code (or putting it in `if__name__ == "__main__"`) from files that are imported elsewhere (see point 4)
   4. commenting code in/out to enable inference could perhaps be improved via a flag (see point 4)
1. Could a common entry point (e.g. similar to https://github.com/jpata/particleflow/blob/master/mlpf/pipeline.py) be defined, from which the various training, inference and validation codes can be run?
2. Could you please make sure the basic pytorch functionality (dataset generation, training, inference) is being exercised in the github CI pipeline, and the pipeline runs successfully?
3. Can you check if `numba` is actually used in the pytorch code, and if it's not used, remove the import (it is a somewhat heavy dependency)?

**Comment by faroukmokhtar** [comment]
Thanks for your suggestions! I removed many unnecessary files to avoid confusion. Here are some answers to the points above:

1) PR description: added a pipeline.py file to train on delphes dataset, make predictions and make evaluation plots. Furthermore, added an LRP machinery that computes relevance scores to explain the model predictions. Check the README.md for more information.

2) Currently the model works perfectly with delphes. It should be straightforward to test if it works with cms, and make simple changes it doesn't). I will work on this when I get back.

3) Worked on most of these. Still have to: 
- add __init__.py as a workaround sys.path.insert
- make gpu detection check in only one file (the main file to run). Some files still need to get updated in this regard. Specifically, the LRP files.

4) Running pipeline.py (the rename of training.py) with different flags will either train a model, make predictions on data, and/or make evaluation plots.

5) All checks have passed now.

6) Removed numba import.

Let me know if you have any more suggestions.

**Comment by jpata** [comment]
Thanks, this is much improved. If the LRP does not work in the test environment:

```
 ./scripts/local_test_delphes_pytorch.sh: line 61:  2244 Killed                  python3 lrp_pipeline.py --n_test=1 --batch_size=4 --lrp_dataset_qcd='../test_tmp_delphes/data/pythia8_qcd' --lrp_outpath='../test_tmp_delphes/experiments/' --lrp_load_model='PFNet7_gen_ntrain_1_nepochs_10_batch_size_4_lr_0.0001_alpha_0.0002_both__nn1_nn3'
Error: Process completed with exit code 137.
```

could you remove it from `scripts/local_test_delphes_pytorch.sh`?

**Comment by faroukmokhtar** [comment]
Done. Now it passed all the github checks.

Let me know if there's anything else to edit/improve to merge the PR.

**Comment by jpata** [comment]
Thanks! Merging this now.

**Pull request merged**